### PR TITLE
fix issues with `{getFrameData,submitFrame} must be called within a VRDisplay.requestAnimationFrame callback.` (fixes issue #65); fix JS errors `Screen position out of view frustum` in Chrome Canary (fixes issue #64)

### LINKED ISSUE #64: JS `Screen position out of view frustum` errors in Chrome Canary
0. ensure you are on Windows and have Steam VR installed (tested on Dell Visor MR and Vive headsets)
1. download **[Chrome Canary](https://www.google.com/chrome/browser/canary.html)** for Windows
1. ensure these Chrome flags are all marked as **`Enabled`**:
    - WebVR — `chrome://flags#enable-webvr`
    - OpenVR — `chrome://flags#openvr`
    - Gamepad Extensions — `chrome://flags#enable-gamepad-extensions`
1. load https://caseyyee.github.io/unity-webvr-export/Build/
1. after "Loading..." screen, the **scene is black**. until I press the "Enter VR" button, there are constant **Console errors**: `Screen position out of view frustum    blob:https://caseyye…c-66cfdf698c02:7636 `
1. when I click the "Enter VR" button, the errors stop and the stereo, mirrored canvas appears as expected. when I toggle from mono to stereo, after that, it's fine.

### LINKED ISSUE #65: VR mode not initially shown in Chrome Canary because of Console warnings: `{getFrameData,submitFrame} must be called within a VRDisplay.requestAnimationFrame callback.`
0. ensure you are on Windows and have Steam VR installed (tested on Dell Visor MR and Vive headsets)
1. load https://caseyyee.github.io/unity-webvr-export/Build/ in [Chrome Canary](https://www.google.com/chrome/browser/canary.html)
2. click the VR-icon button (you'll encounter issue #64)
3. click the VR-icon again (to toggle VR mode)
4. upon exiting, you'll see this Console warning repeatedly: **`getFrameData must be called within a VRDisplay.requestAnimationFrame callback.`**
5. upon re-entering, you'll see this Console warning: **`submitFrame must be called within a VRDisplay.requestAnimationFrame callback.`**

image
image
image
image
image

### PR DESCRIPTION
**Pull request by cvan**
Fix #64 
Fix #65 

seems to work well. feedback to approach, including possible edge cases I've missed, welcome.
tested with HTC VIVE on Windows 10 in both Chrome Canary and Firefox Nightly.

### CODE REVIEW

**Comment by cvan** [comment]
> @cvan Tested with FF and Chrome Canary, looks to be working well. The only thing that broke is that we're starting out in the ground.

thanks for testing. any ideas how to fix? did you check + build on `master` to compare?

File: Build/index.html [hunk 1]
```diff
@@ -3,8 +3,8 @@
   <head>
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width, initial-scale=1.0">
-    <title>Unity-WebVR-Export</title>
-    <meta name="description" content="Unity-WebVR-Export">
+    <title> | Unity-WebVR-Export</title>
```
Comment (cvan) [inline Build/index.html]: I'm surprised this hasn't been updated on `master` when y'all have built the project

File: Assets/WebGLTemplates/WebVR/webvr.js [hunk 2]
```diff
@@ -33,66 +34,98 @@
       });
     }
   }
-  
-  function onUnityLoaded() {
+
+  function onUnityLoaded () {
+    // Get and hide Unity's `<canvas>` instance.
```
Comment (delapuente) [inline Assets/WebGLTemplates/WebVR/webvr.js]: I don't get the comment. What you are hiding is the loader, not the `<canvas>`.

File: Assets/WebGLTemplates/WebVR/webvr.js [hunk 3]
```diff
@@ -33,66 +34,98 @@
...
+      console.error('Unable to enter VR mode:', err);
     });
   }
 
-  function onExitPresent() {
-    vrDisplay.exitPresent().then(function () {
-      console.log('vr exit present success!')
+  function onExitPresent () {
+    if (!vrDisplay && !vrDisplay.isPresenting) {
+      throw new Error('No VR display to exit VR mode');
```
Comment (delapuente) [inline Assets/WebGLTemplates/WebVR/webvr.js]: I would use a `console.warn()` here and fail silently but it's your call.

File: Assets/WebGLTemplates/WebVR/webvr.js [hunk 4]
```diff
@@ -33,66 +34,98 @@
...
+      return;
+    }
 
     if (vrDisplay) {
+      // When this is called for the first time, it will be using the standard
+      // `window.requestAnimationFrame` API, which will throw a Console warning when we call
+      // `vrDisplay.submitFrame(…)`. So for the first frame that this is called, we will
+      // abort early and request a new frame from the VR display instead.
+      if (vrDisplay.isPresenting && !submitNextFrame) {
+        submitNextFrame = false;
```
Comment (delapuente) [inline Assets/WebGLTemplates/WebVR/webvr.js]: If the conditional is `true`, then `submitNextFrame` is already `false`. What's the point of this assignment?

File: Assets/WebGLTemplates/WebVR/webvr.js [hunk 5]
```diff
@@ -33,66 +34,98 @@
...
-    gameInstance.SendMessage('WebVRCameraSet', 'End');
-    onResize();
   }
 
-  function onAnimate() {
-    // RAF has been shimmed, See onRequestAnimationFrame function
-    window.requestAnimationFrame(onAnimate);
+  function onAnimate () {
+    if (!vrDisplay || !vrDisplay.isPresenting) {
+      windowRaf(onAnimate);
```
Comment (delapuente) [inline Assets/WebGLTemplates/WebVR/webvr.js]: I think this is a kind of _active polling_, so you repeat the check on `vrDisplay`. I would prefer this to be event-driven but if that's not possible, can you clarify why you are rescheduling `onAnimate` in a comment?

File: Build/index.html (hunk 1 above)
Comment (delapuente) [reply Build/index.html]: Me too. I don't understand why.

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 5 above)
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/webvr.js]: I could actually remove these lines, good call. they were leftovers.

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 4 above)
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/webvr.js]: doy, that's what I get for working late. I should be checking only `!submitNextFrame`. thanks.

and that order does matter; I don't know if the comment above makes sense. anything I clarify?

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 3 above)
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/webvr.js]: for the `VRManager`, I've got in a local branch (initially started for this, and then for #104, and so on), I have turned these functions into methods, and this places nicely with it.

I'm cool with removing that for now. thanks for mentioning the inconsistency; it was bothering me too

File: Build/index.html (hunk 1 above)
Comment (cvan) [reply Build/index.html]: have issue #116 filed. (I just upgraded to the `.3` version of Unity; mayve that'll change things. I'll let you know.) thanks!

File: Build/manifest.webmanifest [hunk 6]
```diff
@@ -22,7 +22,7 @@
   ],
   "screenshots": [
     {
-      "src": "https://raw.githubusercontent.com/caseyyee/unity-webvr-export/master/img/preview.gif",
+      "src": "https://raw.githubusercontent.com/mozilla/unity-webvr-export/master/img/preview.gif",
```
Comment (cvan) [inline Build/manifest.webmanifest]: @delapuente: how come these changes weren't changed when you did the builds manually yourself? do we have something wonky set up here?

I'd be curious to get on Vidyo together, the three of us, and compare our Unity flows. I'm sure there are some interesting workflows and edge cases we're handling differently.

**Comment by cvan** [comment]
I pulled, rebased, but I'm still getting these conflicts according to GitHub. hmm… I'll have to look in the morning (my morning).

File: Build/manifest.webmanifest (hunk 6 above)
Comment (delapuente) [reply Build/manifest.webmanifest]: Happy to do it. Send an invitation if you want.

**Comment by cvan** [comment]
@delapuente @caseyyee: ignoring the `Build/` mess, I just want to make sure that these changes (really just to `Assets/WebGLTemplates/WebVR/webvr.js` and `Build/webvr.js`) run smoothly for you and with no discernible regressions.

thanks for reviewing and testing, you both :+1:

**Comment by delapuente** [comment]
No problem Chris, if you can not solve the issues, I will rebase it. I'm having problems with my VIVE right now and cannot wake the base stations up so I can not test it.

**Comment by delapuente** [comment]
This is inevitable:
> @cvan Tested with FF and Chrome Canary, looks to be working well. The only thing that broke is that we're starting out in the ground.

Now, when not presenting, we are not updating the camera matrices. For me, it makes perfect sense since we are returning the control to Unity's camera system. If this is the desired behaviour, we must also reset the effect of setting `worldToCameraMatrix` by calling [`Camera#ResetWorldToCameraMatrix.html`](https://docs.unity3d.com/ScriptReference/Camera.ResetWorldToCameraMatrix.html) on all the cameras of the WebVR rigging. It can be done when handling the `End` message sent from JavaScript to Unity.

**Comment by delapuente** [comment]
@cvan working as expected on my side. Solve the issues and grant me access to your repo if you need help with the rebasing. Nice work. These are all valuable steps towards #104

**Comment by caseyyee** [comment]
@delapuente with the polyfill, it should still update (this should be the old version still?)

If we replaced it with the newer one, then yeah.  It won't hit that code path.

**Comment by delapuente** [comment]
> @delapuente with the polyfill, it should still update (this should be the old version still?)

Even with the polyfill, the execution will never hit that code path (notice the [test condition and the early return](https://github.com/cvan/unity-webvr-export/blob/399ccf5c59b2341cde629d91355cb900ab7d25ff/Assets/WebGLTemplates/WebVR/webvr.js#L124-L127))

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 4 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/webvr.js]: Something like `the flag, if true, ensures that we are inside a vrDisplay.requestAnimationFrame`. Anyway, it was clear enough for me. Perhaps we can think of a better fix when addressing #104

**Comment by delapuente** [comment]
@cvan, I've discarded your changes on `Build/` to ease the rebase. I'll regenerate them and push to master once this is merged. Please, address the changes and remember to not commit the changes on `Build`. Thank you so much, let's fix two bugs for the price of one!

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 5 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/webvr.js]: If they are leftovers, please, do it.

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 3 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/webvr.js]: I saw you replaced a different `throw` with a warning but what about this exception?

Comment (cvan) [reply Assets/WebGLTemplates/WebVR/webvr.js]: oh, I missed this one. got distracted with the launch logistics today. will resivit this on Tuesday. (I’m PTO tomorrow, and Monday is a US holiday.)

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 5 above)
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/webvr.js]: I haven’t step-debugged the rAF calls. but I think if I add a `debugger` call to the monkeypatched `window.requestAnimationFrame` function, I can see how often and when Unity’ s WebGL Engine makes the calls. I want to minimise the rAF calls as much as possible. in my testing today, I removed this block and didn’t notice any degradation.

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 3 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/webvr.js]: No problem. I'll do.

File: Assets/WebGLTemplates/WebVR/webvr.js (hunk 5 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/webvr.js]: Non-blocking. Let's move to #104

**Pull request merged**